File: app/src/main/java/com/explorify/xplore/xplore/SearchDestinationActivity.java
package com.explorify.xplore.xplore;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by nikao on 2/23/2017.
 */


public class SearchDestinationActivity extends Activity {

    private static final int CHOSEN_DEST_DEFAULT_VAL = -1;
    private int chosenDestId = CHOSEN_DEST_DEFAULT_VAL;
    private boolean first = true;
    private ListView list;
    private EditText searchBar;
    private List<Integer> resultIDs = new ArrayList<>();
    private ArrayList<ReserveButton> answerButtons = new ArrayList<>();
    private ArrayList<ReserveButton> reserveButtons = new ArrayList<>();

    DBManager dbManager;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.search_layout);

        //setting up the listview
        list = (ListView) findViewById(R.id.resultsListView);

        //TODO convert this to java and skip the other crap arguments
        dbManager = new DBManager(this, "reserveDB.db", General.DB_TABLE);
        dbManager.openDataBase();

        //setting up searchbar
        searchBar = (EditText) findViewById(R.id.searchEditText);
        searchBar.setSingleLine(true);
        searchBar.setHint(R.string.search_hint);
        searchBar.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    searchListItems(charSequence.toString().toLowerCase());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });


    }

    @Override
    protected void onResume() {
        super.onResume();
        populateButtonList(reserveButtons, this);

        if(first) {
            first = false;
            answerButtons.addAll(reserveButtons);
            populateListView();
        }
    }

    //TODO after converting to kotlin, do this asynchronously
    private void populateButtonList(ArrayList<ReserveButton> reserveButtons, Context context)
    {
        String table = General.DB_TABLE;
        Resources resources = context.getResources();

        reserveButtons.clear();

        //Getting each resID separately
        //TODO this is utter shit, put the loop inside of DBManager so it doesn't create and destroy a goddamn cursor every time we need a string from DB
        for(int i = 0; i < MainActivity.RESERVE_NUM; i++)
        {
            //TODO change this with convertFromDrawableNameToId
            int resid = resources.getIdentifier(dbManager.getStr(i, DBManager.ColumnNames.getIMAGE(), table),"drawable","com.explorify.xplore.xplore");
            reserveButtons.add (
                    new ReserveButton(i, ContextCompat.getDrawable(context, resid),
                            dbManager.getStr(i, DBManager.ColumnNames.getNAME(), table))
            );
        }
    }

    private void populateListView() {
        //creating list
        ArrayAdapter<ReserveButton> adapter = new MyListAdapter();
        list.setAdapter(adapter);
    }

    private class MyListAdapter extends ArrayAdapter<ReserveButton> {
        public MyListAdapter() {
            super(SearchDestinationActivity.this, R.layout.list_item, answerButtons);
        }

        @NonNull
        @Override
        public View getView(final int position, View convertView, ViewGroup parent) {
            View itemView = convertView;
            if (itemView == null) {
                itemView = getLayoutInflater().inflate(R.layout.list_item, parent, false);
            }
            final ReserveButton currentButton = answerButtons.get(position);

            Button butt = (Button) itemView.findViewById(R.id.resultItem);
            butt.setText(currentButton.getName());
            butt.setBackground(currentButton.getImage());

            //Configuring Clicks
            butt.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    General.HideKeyboard(SearchDestinationActivity.this);
                    AlertDialog.Builder builder = new AlertDialog.Builder(SearchDestinationActivity.this);
                    builder.setMessage(getResources().getString(R.string.question_choose_reserve)+" "+currentButton.getName()+"?")
                            .setTitle(currentButton.getName())
                            .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialogInterface, int i) {
                                    chosenDestId = currentButton.getId();
                                    onBackPressed();
                                }
                            })
                            .setNegativeButton(R.string.no, null);

                    builder.show();
                }
            });

            return itemView;
        }
    }

    @Override
    public void onBackPressed() {
        Intent resultIntent = new Intent();
        if(chosenDestId == CHOSEN_DEST_DEFAULT_VAL){
            setResult(Activity.RESULT_CANCELED);
        }
        else {
            resultIntent.putExtra("chosen_destination_id", chosenDestId);
            setResult(Activity.RESULT_OK, resultIntent);
        }
        finish();

    }

    private void searchListItems(String query) {
        answerButtons.clear();

        //Searching Database
        resultIDs = dbManager.getIdFromQuery(query, General.DB_TABLE);

        //Returning Results
        if(resultIDs == null)
        {
            Toast.makeText(this, R.string.search_no_results , Toast.LENGTH_SHORT).show();
        }
        else {
            int index = 0;
            for (int result : resultIDs) {
                //result is the single ID of an answer
                answerButtons.add(index,reserveButtons.get(result));
                index ++;
            }
            populateListView();
        }
    }
}

File: app/src/main/java/com/explorify/xplore/xplore/SearchDestinationActivity.kt
package com.explorify.xplore.xplore

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import kotlinx.android.synthetic.main.list_item_reserve.view.*
import kotlinx.android.synthetic.main.search_layout2.*

import java.util.ArrayList

/**
 * Created by nikao on 2/23/2017.
 */


class SearchDestinationActivity : Activity() {

    private val dbManager: DBManager by lazy { DBManager(this) }

    companion object {
        private val CHOSEN_DEST_DEFAULT_VAL = -1
        private var chosenDestId = CHOSEN_DEST_DEFAULT_VAL
    }

    private var reserveCards = ArrayList<ReserveCard>()
    private val answerCards = ArrayList<ReserveCard>()
    private val iconList = arrayListOf<Int>(
            R.drawable.ic_tent_grey600_36dp,
            R.drawable.ic_nature_people_grey600_36dp,
            R.drawable.ic_castle_grey600_36dp,
            R.drawable.ic_paw_grey600_36dp
    )

    private var resultIDs: List<Int> = ArrayList()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.search_layout2)

        init()
        firstDisplayData()
    }

    fun init(){
        //Starting the loading animation
        progressBar.visibility = View.VISIBLE

        //setting up searchbar
        searchEditText.setSingleLine(true)
        searchEditText.setHint(R.string.search_name_hint)
        searchEditText.setSelectAllOnFocus(true)
        searchEditText.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) { }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) { }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                //Log.println(Log.INFO, "brejk", "${s.toString()} is the search query")
                searchListItems(s.toString().toLowerCase())
            }
        })

        //Clear answer list
        answerCards.clear()

        //Setting layoutmanager
        resultsRV.layoutManager = LinearLayoutManager(this)
    }

    fun firstDisplayData() {
        dbManager.openDataBase()

        //Load all reserveCards in a separate thread
        Thread(Runnable {
            //Loading data
            reserveCards = dbManager.getAllReserveCards()
            answerCards.addAll(reserveCards)
            //Creating & setting adapter
            val adapter = RVAdapter(answerCards, this, iconList)
            resultsRV.post { resultsRV.adapter = adapter }
            progressBar.post { progressBar.visibility = View.INVISIBLE }
        }).start()
    }

    private fun searchListItems(query: String) {
        //Resetting answers
        answerCards.clear()

        //Searching Database
        resultIDs = dbManager.getIdFromName(query, General.DB_TABLE)

        //Returning Results
        if (!resultIDs.isEmpty()) {
            var index = 0
            for (result in resultIDs) {
                //result is the single ID of an answer
                answerCards.add(index, reserveCards[result])
                index++
            }
        }
        //Displaying the changes/results
        resultsRV.adapter.notifyDataSetChanged()
    }

    class RVAdapter(val results: List<ReserveCard>, val activity: Activity, val icons: ArrayList<Int>)
        : RecyclerView.Adapter<RVAdapter.ResultViewHolder>(){

        class ResultViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView){
            internal val reserveName: TextView
            internal val reserveImage: ImageView
            internal val reserveIcon: ImageView

            init{
                reserveName = itemView.reserveNameTextView
                reserveImage = itemView.reserveImageView
                reserveIcon = itemView.reserveIconImageView
            }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ResultViewHolder =
                ResultViewHolder(
                        LayoutInflater.from(parent.context)
                                .inflate(R.layout.list_item_reserve, parent, false)
                )

        override fun onBindViewHolder(holder: ResultViewHolder, position: Int) {
            holder.reserveName.setText(results[position].name)
            holder.reserveImage.setImageResource(results[position].imageId)
            holder.reserveIcon.setImageResource(icons[results[position].iconId])
            holder.itemView.setOnClickListener {
                General.HideKeyboard(activity)
                showConfirmDialog(results[position])
            }
        }

        override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
            super.onAttachedToRecyclerView(recyclerView)
        }

        fun showConfirmDialog(reserveCard: ReserveCard) {
            //Building the alert dialog
            val builder = AlertDialog.Builder(activity);
            val message = "${activity.resources.getString(R.string.question_choose_reserve)} ${reserveCard.name}?"
            builder.setMessage(message)
                    .setTitle(reserveCard.name)
                    .setPositiveButton(R.string.yes, object : DialogInterface.OnClickListener {
                        override fun onClick(dialog: DialogInterface?, which: Int) {
                            chosenDestId = reserveCard.id
                            Log.println(Log.INFO, "brejk", "chosen dest id is $chosenDestId")
                            activity.onBackPressed()
                        }
                    })
                    .setNegativeButton(R.string.no, null);

            builder.show();
        }

        override fun getItemCount(): Int = results.size
    }

    override fun onBackPressed() {
        val resultIntent = Intent()
        //Checking if destination was chosen
        if (chosenDestId == CHOSEN_DEST_DEFAULT_VAL) {
            setResult(Activity.RESULT_CANCELED)
        } else {
            resultIntent.putExtra("chosen_destination_id", chosenDestId)
            setResult(Activity.RESULT_OK, resultIntent)
        }
        finish()
    }
}